core/network/src/p2p/gossip/gossip.rs

```diff
@@ -0,0 +1,324 @@
+use crate::p2p::client::Client;
+use crate::p2p::gossip::{Gossipable, Set};
+use crate::p2p::sdk::{PullGossipRequest, PullGossipResponse};
+use avalanche_types::ids::Id;
+use log::{debug, error};
+use probabilistic_collections::bloom::BloomFilter;
+use prost::Message;
+use std::error::Error;
+use std::hash::Hash;
+use std::marker::PhantomData;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tokio::select;
+use tokio::sync::mpsc::{channel, Receiver};
+use tokio::time::interval;
+
+pub struct Config {
+    pub namespace: String,
+    pub frequency: Duration,
+    pub poll_size: usize,
+}
+
+pub struct Gossiper<T: Gossipable, S: Set<T>> {
+    config: Config,
+    set: Arc<Mutex<S>>,
+    client: Arc<Client>,
+    stop_rx: Receiver<()>,
+    phantom: PhantomData<T>, // Had to use this to please the compiler about T not being used.
```

Comment:
Yea was not happy with that as well while implementing it. I found another way to go around this issue, using associated type (which I did not know before hand TIL i guess)

Reply:
Quick lesson on when to use associated types vs. generics (in case you don't already know this):

A generic type represents many types, right? So, would it make sense to have many _different_ trait implementations, one for each concrete type? OR... is there a single type that should be _associated_ with this particular trait implementation?

Say the `Iterator` trait did not exist, and we wanted to create an `Iterator` trait and implement it for `Vec<T>` (not actually how this works, but I think the example will be clear).  Would we want to have a different implementation for a `Vec<String>` vs. a `Vec<u8>`? Definitely not! Here's where we want an _associated_ type. 

A good rule of thumb is to actually start with an associated type and see if you can get it to work the way you want; it will become clear very quickly if you actually need to make your trait generic.